pkg/dns/config/sync_dir_test.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"reflect"
+	"testing"
+	"time"
+
+	"os"
+
+	"k8s.io/client-go/pkg/util/clock"
+)
+
+func TestSyncFile(t *testing.T) {
+
+	testParentDir, err := ioutil.TempDir("", "test.filesyncsource")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() { os.RemoveAll(testParentDir) }()
+
+	testDir := filepath.Join(testParentDir, "datadir")
+
+	fakeClock := clock.NewFakeClock(time.Now())
+
+	source := newFileSyncSource(testDir, time.Second, fakeClock)
+
+	// missing dir should error
+	if _, err := source.Once(); err == nil {
+		t.Fatalf("expected error reading missing dir")
+	}
+
+	// empty dir should return empty results
+	if err := os.Mkdir(testDir, os.FileMode(0755)); err != nil {
+		t.Fatal(err)
+	}
+	result, err := source.Once()
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if result.Version != "" || len(result.Data) != 0 {
+		t.Fatalf("expected empty version and data reading empty dir, got %#v", result)
+	}
+
+	// should not recurse and should ignore dot files
+	if err := os.Mkdir(filepath.Join(testDir, "subdir"), os.FileMode(0755)); err != nil {
+		t.Fatal(err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(testDir, "subdir", "subdirfile"), []byte("test"), os.FileMode(0755)); err != nil {
+		t.Fatal(err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(testDir, ".hiddenfile"), []byte("test"), os.FileMode(0755)); err != nil {
+		t.Fatal(err)
+	}
+	result, err = source.Once()
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if result.Version != "" || len(result.Data) != 0 {
+		t.Fatalf("expected empty version and data reading dir containing subdirs and dotfiles, got %#v", result)
+	}
+
+	// should return error if non-utf8 data is encountered
+	// https://en.wikipedia.org/wiki/UTF-8#Codepage_layout
+	if err := ioutil.WriteFile(filepath.Join(testDir, "binary"), []byte{192}, os.FileMode(0755)); err != nil {
```

Comment:
it's `map[string]string`, not `map[string][]byte`, so it should only be utf-8 data